test/js-api/gc/casts.tentative.any.js

```diff
@@ -0,0 +1,200 @@
+  const structIndex = builder.addStruct([makeField(kWasmI32, true)]);
+  const arrayIndex = builder.addArray(kWasmI32, true);
+  const structIndex2 = builder.addStruct([makeField(kWasmF32, true)]);
+  const arrayIndex2 = builder.addArray(kWasmF32, true);
+  const funcIndex = builder.addType({ params: [], results: [] });
+
+  const anyRef = wasmRefType(kWasmAnyRef);
+  const eqRef = wasmRefType(kWasmEqRef);
+  const structRef = wasmRefType(kWasmStructRef);
+  const arrayRef = wasmRefType(kWasmArrayRef);
+  const i31Ref = wasmRefType(kWasmI31Ref);
+  const funcRef = wasmRefType(kWasmFuncRef);
+
+  builder
+    .addFunction("anyArg", makeSig_v_x(anyRef))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("eqArg", makeSig_v_x(eqRef))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("structArg", makeSig_v_x(structRef))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("arrayArg", makeSig_v_x(arrayRef))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("i31Arg", makeSig_v_x(i31Ref))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("funcArg", makeSig_v_x(funcRef))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("concreteStructArg", makeSig_v_x(wasmRefType(structIndex)))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("concreteArrayArg", makeSig_v_x(wasmRefType(arrayIndex)))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("concreteFuncArg", makeSig_v_x(wasmRefType(funcIndex)))
+    .addBody([])
+    .exportFunc();
+
+  builder
+    .addFunction("makeStruct", makeSig_r_v(wasmRefType(structIndex)))
+    .addBody([...wasmI32Const(42),
+              ...GCInstr(kExprStructNew), structIndex])
+    .exportFunc();
+
+  builder
+    .addFunction("makeArray", makeSig_r_v(wasmRefType(arrayIndex)))
+    .addBody([...wasmI32Const(5), ...wasmI32Const(42),
+              ...GCInstr(kExprArrayNew), arrayIndex])
+    .exportFunc();
+
+  builder
+    .addFunction("makeStruct2", makeSig_r_v(wasmRefType(structIndex2)))
+    .addBody([...wasmF32Const(42),
+              ...GCInstr(kExprStructNew), structIndex2])
+    .exportFunc();
+
+  builder
+    .addFunction("makeArray2", makeSig_r_v(wasmRefType(arrayIndex2)))
+    .addBody([...wasmF32Const(42), ...wasmI32Const(5),
+              ...GCInstr(kExprArrayNew), arrayIndex2])
+    .exportFunc();
+
+  builder
+    .addFunction("testFunc", funcIndex)
+    .addBody([])
+    .exportFunc();
+
+  const buffer = builder.toBuffer();
+  const module = new WebAssembly.Module(buffer);
+  const instance = new WebAssembly.Instance(module, {});
+  exports = instance.exports;
+});
+
+test(() => {
+  exports.anyArg(exports.makeStruct());
+  exports.anyArg(exports.makeArray());
+  exports.anyArg(42);
+  exports.anyArg(42n);
+  exports.anyArg("foo");
+  exports.anyArg({});
+  exports.anyArg(() => {});
+  assert_throws_js(TypeError, () => exports.anyArg(exports.testFunc));
```

Comment:
This line of the anyref test fails on V8 and SpiderMonkey because they do not throw. If I'm reading the JS API spec correctly though, `ToWebAssemblyValue` should throw here.
 
The relevant steps of the algorithm are from step 6:

> 6. If type is of the form [ref](https://webassembly.github.io/spec/core/syntax/types.html#syntax-reftype) null heaptype,
> [...]
>    4. Else if v is an [Exported Function](https://webassembly.github.io/gc/js-api/index.html#exported-function),
>      1. Let funcaddr be the value of v’s [[FunctionAddress]] internal slot.
>      2. Let r be [ref.func](https://webassembly.github.io/spec/core/exec/runtime.html#values) funcaddr.

In this case, `exports.testFunc` should be an exported function and this step will match. Thus `r` will be a `ref.func <addr>` and its type will be a typed funcref.

> 8. Let actualtype be [ref_type](https://webassembly.github.io/spec/core/appendix/embedding.html#embed-ref-type)(store, r).
> 9. If [match_valtype](https://webassembly.github.io/spec/core/appendix/embedding.html#embed-match-valtype)(actualtype, type) is false,
>    1. Throw a [TypeError](https://tc39.github.io/ecma262/#sec-native-error-types-used-in-this-standard-typeerror).

Here `actualtype` should be some function type (is it specified that `ref_type` must return the most specific type?). `match_type(<functype>, (ref any))` will return false because functions cannot be a subtype of any. Thus this should throw.

Does this seem like the correct reading of the spec?

Reply:
Yes, that's the correct reading. Whether it's actually doing what's intended is a different question. :)  Given that any JS value is in externref, and every externref can be converted to anyref, we may want to allow all JS values for any. For that, we'd need to add a special case under step 9, that if heaptype is any, instead of throwing, construct a ref.host (as in step 6).

(Yes, ref_type returns the most specific type.)

Reply:
Ok, that makes sense to me given that wasm's type system will keep exported functions as hostrefs and funcrefs separate. I can change this test to not throw, in which case the browsers will all agree (for JSC the patch for this behavior hasn't landed, so I can just change it).

For the special case, it seems like it would be equivalent to say in step 4:

> Else if v is an [Exported Function](https://webassembly.github.io/gc/js-api/index.html#exported-function) and heaptype != any

in which case it falls through to the hostref case (so that it's not necessary to repeat the ref.host construction prose).

Reply:
Sounds good, please see #440.

Reply:
I just updated this test case to match the new spec PR.